designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,312 @@
+# Disruption Controls By Reason
+## TOC and Overview
+1. [Overview](#overview)
+2. [User Scenarios](#user-scenarios)
+3. [Known Requirements](#clarifying-the-requirements-and-behavior)
+   - [Q: Default or Undefined Reason Case Handling](#q-default-or-undefined-reason-case-handling)
+   - [Q: Calculation of Allowed Disruptions in a Multi-Reason World](#q-calculation-of-allowed-disruptions-in-a-multi-reason-world)
+   - [Q: Handling an Unspecified Default Reason](#q-handling-an-unspecified-default-reason)
+   - [Q: Definition of Default in Budgets](#q-definition-of-default-in-budgets)
+4. [API Design](#api-design)
+   - [Approach A: Extending the Budget API to Specify a Reason](#approach-a-extending-the-budget-api-to-specify-a-reason)
+      * [List Approach: Multiple Reasons per Budget](#list-approach-multiple-reasons-per-budget)
+      * [Single Reason Approach: One Reason per Budget](#single-reason-approach-one-reason-per-budget)
+      * [Pros and Cons for Both Approaches](#pros-and-cons-for-both-approaches)
+      * [Preferred Option: List Approach](#preferred-option-list-approach)
+   - [Approach B: Defining Per Reason Controls](#approach-b-defining-per-reason-controls)
+      * [Pros and Cons](#pros-and-cons)
+   - [API Design Conclusion: Preferred Design](#api-design-conclusion-preferred-design)
+
+## User Scenarios 
+1. Users need the capability to schedule upgrades only during business hours or within more restricted time windows. Additionally, they require a system that doesn't compromise the cost savings from consolidation when upgrades are blocked due to drift.
+2. Users want to minimize workload disruptions during business hours but still want to be able to delete empty nodes throughout the day.  That is, empty can run all day, while limiting cost savings and upgrades due to drift to non-business hours only.
+
+See Less Made Up Scenarios here: 
+- https://github.com/kubernetes-sigs/karpenter/issues/924 
+- https://github.com/kubernetes-sigs/karpenter/issues/753#issuecomment-1790110838
+- https://github.com/kubernetes-sigs/karpenter/issues/672
+
+## Clarifying the requirements and behavior 
+**Reason and Budget Definition:** Users should be able to define an reason and a corresponding budget(s).
+**Supported Reasons:** All disruption Reasons affected by the current Budgets implementation (underutilized, empty, expired, drifted) should be supported. 
+**Default Behavior for Unspecified Reasons:** Budgets should continue to support a default behavior for all disruption reasons. 
+
+
+### Q: How should Karpenter handle the default or undefined reason case? 
+If a budget reason is unspecified like budgets[2], we will assume this budget applies to all actions that are not specified 
```

Comment:
I can't tell if this is an artifact of a previous version or you decided to 1-index this :D 
```suggestion
If a budget reason is unspecified like budgets[1], we will assume this budget applies to all actions that are not specified 
```

Reply:
Yeah this doc went through too many adhoc adjustments. Need to go through and properly clean up.